.ci/Jenkinsfile

```diff
@@ -0,0 +1,89 @@
+
+// Verified with Jenkins v2.164.3
+
+// TODO:
+// 1. Calculate taskset/affinity for the scripts based on total number of jenkins executors
+
+pipeline {
+    agent {label "ml-test-node-gpu"}
+
+    options {
+        buildDiscarder(logRotator(numToKeepStr:'10'))
+        timeout(time: 15, unit: 'MINUTES')
+        disableConcurrentBuilds()
+    }
+
+    environment {
+        ARTEFACT_DIR = "${WORKSPACE}/artefacts"
+        NFS_WORKSPACE = "${NFS_WORKSPACE_ROOT}/ml-nccl-rdma-sharp-plugins-pr/${BUILD_NUMBER}"
+        NCCL_SRC_DIR = "${WORKSPACE}/nccl"
+        NCCL_DIR = "${NFS_WORKSPACE}/nccl"
+        NCCL_TESTS_DIR = "${NFS_WORKSPACE}/nccl-tests"
+    }
+
+    stages {
+        stage('Preparations') {
+            steps {
+                echo 'Preparations...'
+                sh 'mkdir -p ${ARTEFACT_DIR}'
+                sh 'mkdir -p ${NFS_WORKSPACE}'
+            }
+        }
+        stage('Checkout NCCL') {
+            steps {
+                sh 'mkdir -p ${NCCL_SRC_DIR}'
+                dir("${NCCL_SRC_DIR}") {
+                    git branch: 'master',
+                    url: 'git@gitlab.com:Mellanox/nccl-sharp.git'
+                }
+            }
+        }
+        stage('Build NCCL') {
+            steps {
+                echo 'Building NCCL...'
+                sh 'set -o pipefail; ${WORKSPACE}/.ci/build_nccl.sh 2>&1 | tee ${ARTEFACT_DIR}/build_nccl.log'
+                sh 'mkdir -p ${NCCL_DIR}'
+                sh 'cp -R ${NCCL_SRC_DIR}/build/* ${NCCL_DIR}'
+                sh 'find ${NCCL_DIR} -type f'
+            }
+        }
+        stage('Build nccl-rdma-sharp-plugins') {
+            steps {
+                echo 'Building nccl-rdma-sharp-plugins...'
+                sh 'set -o pipefail; ${WORKSPACE}/.ci/build.sh 2>&1 | tee ${ARTEFACT_DIR}/build.log'
+            }
+        }
+        stage('Checkout NCCL tests') {
+            steps {
+                sh 'mkdir -p ${NCCL_TESTS_DIR}'
+                dir("${NCCL_TESTS_DIR}") {
+                    git branch: 'master',
+                    url: 'https://github.com/NVIDIA/nccl-tests.git'
+                }
+            }
+        }
+        stage('Test') {
+            steps {
+                echo 'Testing nccl-rdma-sharp-plugins...'
+                sh 'set -o pipefail; ${WORKSPACE}/.ci/run_nccl_test.sh 2>&1 | tee ${ARTEFACT_DIR}/run_nccl_test.log'
+            }
+        }
+    }
+    post {
+        always {
+            echo 'Post-actions...'
+            echo 'Publish artefacts...'
+            sh 'ls -al ${ARTEFACT_DIR}'
+            sh '''#!/bin/bash -l
+                module load ml/ci-tools
+                env \
+                    UPSTREAM_JOB_NAME=${JOB_NAME} \
+                    UPSTREAM_BUILD_NUMBER=${BUILD_NUMBER} \
+                    UPSTREAM_ghprbGhRepository=${ghprbGhRepository} \
+                    UPSTREAM_ghprbPullId=${ghprbPullId} \
+                    publish_artefacts_to_gist.py
+               '''
+        }
```

Comment:
why not to create script ```publish_artefacts.sh``` with lines 78-85 to externalize logic into script and call it from pipeline?

Reply:
Fixed.